flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamOperatorWrapper.java

```diff
@@ -0,0 +1,217 @@
+	 * {@link #next} points to.
+	 *
+	 * <p>Note that this method must be called in the task thread, because we need to call
+	 * {@link MailboxExecutor#yield()} to take the mails of closing operator and running timers and
+	 * run them.
+	 */
+	public void close(StreamTaskActionExecutor actionExecutor) throws Exception {
+		close(actionExecutor, false);
+	}
+
+	/**
+	 * Ends an input of the operator contained by this wrapper.
+	 *
+	 * @param inputId the input ID starts from 1 which indicates the first input.
+	 */
+	public void endOperatorInput(int inputId) throws Exception {
+		if (wrapped instanceof BoundedOneInput) {
+			((BoundedOneInput) wrapped).endInput();
+		} else if (wrapped instanceof BoundedMultiInput) {
+			((BoundedMultiInput) wrapped).endInput(inputId);
+		}
+	}
+
+	public OP getStreamOperator() {
+		return wrapped;
+	}
+
+	void setPrevious(StreamOperatorWrapper previous) {
+		this.previous = previous;
+	}
+
+	void setNext(StreamOperatorWrapper next) {
+		this.next = next;
+	}
+
+	private void close(StreamTaskActionExecutor actionExecutor, boolean invokingEndInput) throws Exception {
+		if (invokingEndInput) {
+			// NOTE: This only do for the case where the operator is one-input operator. At present,
+			// any non-head operator on the operator chain is one-input operator.
+			actionExecutor.runThrowing(() -> endOperatorInput(1));
+		}
+
+		closeOperatorAndQuiesceTimeService(actionExecutor);
+
+		// propagate the close operation to the next wrapper
+		if (next != null) {
+			next.close(actionExecutor, true);
+		}
+	}
+
+	private void closeOperatorAndQuiesceTimeService(StreamTaskActionExecutor actionExecutor)
+		throws InterruptedException, ExecutionException {
+
+		final CompletableFuture<Void> closedFuture = new CompletableFuture<>();
+
+		// 1. executing the close operation must be deferred to the mailbox to ensure that mails already
+		//    in the mailbox are finished before closing the operator
+		// 2. to ensure that there is no longer output triggered by the timers before invoke the "endInput"
+		//    methods of downstream operators in the operator chain, we must quiesce the processing time
+		//    service to prevent the pending timers from firing, but allow the timers in running to finish
+		// 3. when the second step is finished, send a closed mail to ensure that the timers yet in mailbox
+		//    are finished before exiting the following mailbox processing loop
+		// TODO: To ensure the strict semantics of "close", the operator should be allowed to decide how to
+		//  handle (cancel or trigger) the pending timers before being closed, and the second step should be
+		//  finished before the first step
+		deferCloseOperatorToMailbox(actionExecutor)
+			.thenCompose(unused  -> quiesceProcessingTimeService())
+			.thenRun(() -> sendClosedMail(closedFuture))
+			.exceptionally((throwable) -> forwardException(throwable, closedFuture));
```

Comment:
Won't the `.thenCompose()` be executed only after the future from `deferCloseOperatorToMailbox()` is completed? In other words, operators are quiesced AFTER being closed?

That's different order to what was supposed to happen: https://issues.apache.org/jira/browse/FLINK-14228

has something changed since the ticket was written?

Reply:
> In other words, operators are quiesced AFTER being closed?

Yes. The current call order only ensures the strict semantics of `endInput()`. As mentioned in TODO, the strict semantics of `close()` strictly want to be done in the next step (I started a discussion: http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Deal-with-the-timers-better-before-closing-operators-td36038.html ). If we put `quiesce()` before `close()`, now I only find the problem of `AsyncWaitOperatorTest#testClosingWithBlockedEmitter` failure, which needs to be fixed. What do you think? @pnowojski
```
		// TODO: To ensure the strict semantics of "close", the operator should be allowed to decide how to
		//  handle (cancel or trigger) the pending timers before being closed, and the second step should be
		//  finished before the first step
		deferCloseOperatorToMailbox(actionExecutor)
			.thenCompose(unused  -> quiesceProcessingTimeService())
			.thenRun(() -> sendClosedMail(closedFuture))
			.exceptionally((throwable) -> forwardException(throwable, closedFuture));
```

Reply:
Currently, `AsyncWaitOperator` in close waits for all elements to be either completed or timed out.

The failing test relies on the latter and therefore fails when we disable timers before closing. 
It also fails in the master if we just set operator timeout to a value greater than the test timeout.

So I'd propose a hotfix commit which completes the `resultFuture` in the end of `AsyncWaitOperatorTest.EmitterBlockingFunction#asyncInvoke`.

Reply:
OK. I'll try to change the order to the original propose to see if there are any other problems.